sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala

```diff
@@ -435,11 +435,18 @@ private[sql] case class ParquetRelation2(
     // Push down filters when possible. Notice that not all filters can be converted to Parquet
     // filter predicate. Here we try to convert each individual predicate and only collect those
     // convertible ones.
-    predicates
-      .flatMap(ParquetFilters.createFilter)
-      .reduceOption(FilterApi.and)
-      .filter(_ => sqlContext.conf.parquetFilterPushDown)
-      .foreach(ParquetInputFormat.setFilterPredicate(jobConf, _))
+    if (sqlContext.conf.parquetFilterPushDown) {
+      predicates
+        // Don't push down predicates which reference partition columns
+        .filter { pred =>
+          val partitionColNames = partitionColumns.map(_.name).toSet
+          val referencedColNames = pred.references.map(_.name).toSet
+          referencedColNames.intersect(partitionColNames).isEmpty
+        }
+        .flatMap(ParquetFilters.createFilter)
```

Comment:
I think you guys are optimizing for the wrong thing here.  `reduceOption(<and>)` is an incredibly common pattern when you are munging predicates.  It occurs 15+ times in the code base.

In my experience, `reduce` is always a red flag and it means you aren't considering what to do in the case that the list is empty.  There have been several bugs in catalyst where this was the problem.

You are correct, that you _can_ add an explicit if check.  However, now as I scan the code, I have to connect these two disparate lines and reason about them independently.  When a developer uses `reduceOption` I know that the compiler is going to yell at them if they aren't handling all cases correctly, so I can focus my review on the other logic.